test/e2e/framework/timer/timer.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package timer
+
+import (
+	"time"
+
+	"bytes"
+	"fmt"
+
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/perftype"
+)
+
+var now = time.Now
+
+// Represents a phase of a test. Phases can overlap.
+type Phase struct {
+	sequenceNumber int
+	name           string
+	startTime      time.Time
+	endTime        time.Time
+}
+
+func (phase *Phase) ended() bool {
+	return !phase.endTime.IsZero()
+}
+
+// End marks the phase as ended, unless it had already been ended before.
+func (phase *Phase) End() {
+	if !phase.ended() {
+		phase.endTime = now()
+	}
+}
+
+func (phase *Phase) label() string {
+	return fmt.Sprintf("%03d-%s", phase.sequenceNumber, phase.name)
+}
+
+func (phase *Phase) duration() time.Duration {
+	endTime := phase.endTime
+	if !phase.ended() {
+		endTime = now()
+	}
+	return endTime.Sub(phase.startTime)
+}
+
+func (phase *Phase) humanReadable() string {
+	if phase.ended() {
+		return fmt.Sprintf("Phase %s: %v\n", phase.label(), phase.duration())
+	} else {
+		return fmt.Sprintf("Phase %s: %v so far\n", phase.label(), phase.duration())
+	}
+}
+
+// A TestPhaseTimer groups phases and provides a way to export their measurements as JSON or human-readable text.
+type TestPhaseTimer struct {
+	prep     Phase
```

Comment:
You need to use mutex locking - this is not thread-safe currently (i.e new phases can be created simultaneously).

Reply:
Done.